sway/commands/resize.c

```diff
@@ -62,225 +62,230 @@ static bool resize_floating(int amount, bool use_width) {
+			bool is_after  = use_width ? sibling->x > focused->x : sibling->y > focused->y;
+			if (is_before || is_after) {
+				pixels = -pixels;
+				pixels /= is_before ? nb_before : nb_after;
+				if (nb_after != 0 && nb_before != 0) {
+					pixels /= 2;
+				}
+				sway_log(L_DEBUG, "%p: %s", sibling, is_before ? "before" : "after");
+				if (use_major) {
+					for (int j = i; j < next_i; ++j) {
+						recursive_resize(parent->children->items[j], pixels,
+								 use_width ?
+								 (is_before ? WLC_RESIZE_EDGE_RIGHT : WLC_RESIZE_EDGE_LEFT) :
+								 (is_before ? WLC_RESIZE_EDGE_BOTTOM : WLC_RESIZE_EDGE_TOP));
 					}
+				} else {
+					recursive_resize(sibling, pixels,
+							use_width ?
+							(is_before ? WLC_RESIZE_EDGE_RIGHT : WLC_RESIZE_EDGE_LEFT) :
+							(is_before ? WLC_RESIZE_EDGE_BOTTOM : WLC_RESIZE_EDGE_TOP));
 				}
 			} else {
-				double pixels = amount;
-				if (parent->height + pixels < min_sane_h) {
-					valid = false;
-					break;
-				}
-			}
-		}
-		if (valid) {
-			for (i = 0; i < parent->parent->children->length; i++) {
-				sibling = parent->parent->children->items[i];
-				if (sibling->y != focused->y) {
-					if (sibling->y < parent->y) {
-						double pixels = -1 * amount;
-						pixels /= bnumber;
-						if (tnumber) {
-							recursive_resize(sibling, pixels/2, WLC_RESIZE_EDGE_BOTTOM);
-						} else {
-							recursive_resize(sibling, pixels, WLC_RESIZE_EDGE_BOTTOM);
-						}
-					} else if (sibling->x > parent->x) {
-						double pixels = -1 * amount;
-						pixels /= tnumber;
-						if (bnumber) {
-							recursive_resize(sibling, pixels/2, WLC_RESIZE_EDGE_TOP);
-						} else {
-							recursive_resize(sibling, pixels, WLC_RESIZE_EDGE_TOP);
-						}
+				if (use_major) {
+					for (int j = i; j < next_i; ++j) {
+						recursive_resize(parent->children->items[j], pixels,
+								 use_width ? WLC_RESIZE_EDGE_LEFT : WLC_RESIZE_EDGE_TOP);
+						recursive_resize(parent->children->items[j], pixels,
+								 use_width ? WLC_RESIZE_EDGE_RIGHT : WLC_RESIZE_EDGE_BOTTOM);
 					}
 				} else {
-					if (bnumber != 0 && tnumber != 0) {
-						double pixels = amount/2;
-						recursive_resize(parent, pixels, WLC_RESIZE_EDGE_TOP);
-						recursive_resize(parent, pixels, WLC_RESIZE_EDGE_BOTTOM);
-					} else if (tnumber) {
-						recursive_resize(parent, amount, WLC_RESIZE_EDGE_TOP);
-					} else if (bnumber) {
-						recursive_resize(parent, amount, WLC_RESIZE_EDGE_BOTTOM);
-					}
+					recursive_resize(sibling, pixels,
+							 use_width ? WLC_RESIZE_EDGE_LEFT : WLC_RESIZE_EDGE_TOP);
+					recursive_resize(sibling, pixels,
+							 use_width ? WLC_RESIZE_EDGE_RIGHT : WLC_RESIZE_EDGE_BOTTOM);
 				}
 			}
-			arrange_windows(swayc_active_workspace(), -1, -1);
+			i = next_i;
 		}
-		return true;
+		// Recursive resize does not handle positions, let arrange_windows
+		// take care of that.
+		arrange_windows(swayc_active_workspace(), -1, -1);
```

Comment:
This whole function has now become pretty complicated. Can you split it up into several?